docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,234 @@
+variants. This will be explained in more detail below.
+
+This proposal also introduces a way to "fan-out", or create multiple
+`PackageVariant` resources declaratively based upon a list or selector with the
+`PackageVariantSet` resource.. This is combined with the injection mechanism to
+enable generation of large sets of variants that are specialized to a particular
+target repository, cluster, or other target.
+
+## Basic Package Cloning
+
+The `PackageVariant` resource controls the creation and lifecycle of a variant
+of a package. That is, it defines the original (upstream) package, the new
+(downstream) package, and the changes (mutations) that need to be made to
+transform the upstream into the downstream. It also allows the user to specify
+policies around adoption, deletion, and update of package revisions that are
+under the control of the package variant controller.
+
+![Basic Package Variant](PackageVariant%20-%201.png)
+
+Note that *proposal* and *approval* are not handled by the package variant
+controller. Those are left to humans or other controllers. The exception is the
+proposal of deletion (there is no concept of a "Draft" deletion), which the
+package variant control will do, depending upon the specified deletion policy.
+
+### Package Revision Metadata
+
+The package variant controller utilizes Porch APIs. This means that it is not
+just doing a `clone` operation, but in fact creating a Porch `PackageRevision`
+resource. In particular, that resource can contain Kubernetes metadata that is
+not part of the package as stored in the repository.
+
+Some of that metadata is necessary for the management of the PackageRevision by
+the PackageVariant controller - for example, the `ownerRef` indicating which
+`PackageVariant` created the `PackageRevision`. These are not in control of the
+user. However, the `PackageVariant` resource does make the annotations and
+labels of the `PackageRevision` available as values that may be controlled
+during the creation of the `PackageRevision`. This can assist in additional
+automation workflows.
+
+## Introducing Variance
+Just cloning is not that interesting, so the `PackageVariant` resource also
+allows you to control various ways of mutating the original package to create
+the variant.
+
+### Package Context
+[Note: I believe right now we *always* produce `--for-deployment` package
+revisions out of package variant controller. We will likely need to allows some
+control over that, as we already have a use case in Nephio to use a
+PackageVariant to create a package intended as a base for fan out]
+
+Every kpt package that is designated `--for-deployment` will contain a
+ConfigMap called `kptfile.kpt.dev`. Kpt (or Porch) will automatically add a
+key `name` to the ConfigMap data, with the value of the package name. This
+ConfigMap can then be used as input to functions in the Kpt function pipeline.
+
+This process holds true for package revisions created via the package variant
```

Comment:
I think this could be particularly relevant when combined with the package pipeline and the variant constructor pattern: https://kpt.dev/guides/variant-constructor-pattern.

We have support in PackageVariantSet for creating packages based on other objects referenced by the PackageVariantSet CR. These objects can then provide the context for each variant needed and therefore will be included as parameters in the PackageVariant CRs created by the PackageVariantSet controller. The pipeline included in the packages will then know how to use the context to perform the necessary customizations. It should make it possible to encapsulate the information about customizations and the context in only the targeted objects and the packages themselves.

Reply:
We have injection for that - to insert those target objects and/or other objects. It's described below. But your point is well taken that we may want to make it super convenient to inject the targeting resource.

As for the package context ConfigMap described here, the problem is that it is only created if the downstream repository is flagged as a deployment repository in Porch. That is, the equivalent of the `--for-deployment` flag  value is determined by an attribute of the downstream repo. So, if the downstream repo is not tagged as a deployment repo, I guess we should error out here, if these values have been set?

Reply:
Updated to reflect this.